### PR TITLE
catalog/lease: add migration for session based leasing

### PR DESCRIPTION
**Pull request by fqazi**
This PR will do the following:

1. Add the version numbers required to add session based leasing, and flip through internal states based on these versions.
2. Implement the transitions for these version numbers so that session based leasing is safely adopted (i.e. on going workloads and schema changes are not impacted or waited on properly)
3. Optimize the scanning of leases table for the new session based format

Informs: #119460
informs: https://github.com/cockroachdb/cockroach/issues/116271

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119041)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @fqazi)

___
*[`pkg/cli/zip_table_registry.go` line 1104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYe6II7M717sPGlDGp:-NqYe6II7M717sPGlDGq:btk29cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/cli/zip_table_registry.go#L1104)):*
> ```Go
> 	"system.lease": {
> 		customQueryRedacted: `
> 			SELECT * FROM system.lease;
> ```

don't we want `customQueryUnredacted` here? nothing in the table should be sensitive data
___
*[`pkg/sql/catalog/lease/count.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYez2U6ISLrdg4T2dw:-NqYez2U6ISLrdg4T2dx:b-m1tfj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/sql/catalog/lease/count.go#L44)):*
> ```Go
> 	forAnyVersion bool,
> ) (int, error) {
> 	useSyntheticDescriptors := useSyntheticLeaseDescriptor(settings)
> ```

`useSyntheticLeaseDescriptor` is a bit confusing, since below we are adding to the syntheticDescriptors slice even when `useSyntheticLeaseDescriptor` is false.

maybe we should name this bool and the function something like `useSessionBasedLeaseDescriptor`
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYl2k29RVaFMJU-bZG:-NqYl2k29RVaFMJU-bZH:b-8n38fu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a80f077f42e57cb9a0a5afd31a5b707ba1d7b/pkg/sql/catalog/lease/descriptor_version_state.go#L159)):*
> ```Go
> 	expiration := s.mu.expiration
> 	if s.mu.session != nil &&
> 		s.t.m.sessionBasedLeasingModeAtLeast(context.TODO(), SessionBasedDrain) {
> ```

how bad is it gonna be to get the ctx plumbed through correctly?
___
*[`pkg/sql/catalog/lease/lease.go` line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYToVJ8vzsxqB5pSzj:-NqYToVKBQqKU1bnR2lh:bpkzmfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e608c0a59d612a122f2fc911fa2d0ae60e230aac/pkg/sql/catalog/lease/lease.go#L106)):*
> ```Go
> 	"enables session based leasing for internal testing.",
> 	util.ConstantWithMetamorphicTestChoice("experimental_use_session_based_leasing",
> 		"auto").(string),
> ```

nit: we don't need this to be metamorphic now that it's only going to pick from one choice, right?
___
*[`pkg/sql/catalog/lease/lease.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYeqyL6yo2ZH7MpEdC:-NqYeqyL6yo2ZH7MpEdD:bt7vo4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/sql/catalog/lease/lease.go#L125)):*
> ```Go
> }
> 
> func useSyntheticLeaseDescriptor(settings *cluster.Settings) bool {
> ```

nit: the ctx argument can be added, looks like it's only used in one place, and we have the ctx
___
*[`pkg/sql/catalog/lease/lease.go` line 1638 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYguVp1Nr4oXdZptDK:-NqYguVp1Nr4oXdZptDL:b-4lf9f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/sql/catalog/lease/lease.go#L1638)):*
> ```Go
> 	_ = m.stopper.RunAsyncTask(newCtx, "del-orphaned-leases", func(ctx context.Context) {
> 		// This could have been implemented using DELETE WHERE, but DELETE WHERE
> 		// doesn't implement AS OF SYSTEM TIME.
> ```

nit: the comment seems old now. should we switch to using DELETE WHERE, since we aren't using AOST now?
___
*[`pkg/upgrade/upgrades/upgrades.go` line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYkYszE9QNTeU-UZ8w:-NqYkYt-AXDpA_IeAeQh:byz73b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e608c0a59d612a122f2fc911fa2d0ae60e230aac/pkg/upgrade/upgrades/upgrades.go#L181)):*
> ```Go
> 	),
> 	upgrade.NewTenantUpgrade(
> 		"enable dual writes to system.lease table",
> ```

nit: mention that it mimics `experimental_use_session_based_leasing=dual_write`
___
*[`pkg/upgrade/upgrades/upgrades.go` line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYkdkg4CykfgeRHHFp:-NqYkdkg4CykfgeRHHFq:b3nf3ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/e608c0a59d612a122f2fc911fa2d0ae60e230aac/pkg/upgrade/upgrades/upgrades.go#L189)):*
> ```Go
> 
> 	upgrade.NewTenantUpgrade(
> 		"stop writing expiration based leases to system.lease table",
> ```

nit: mention that it mimics `experimental_use_session_based_leasing=drain`
___
*[`pkg/upgrade/upgrades/upgrades.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYkguhDpPWW0afsN4-:-NqYkguhDpPWW0afsN40:b-69j9j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e608c0a59d612a122f2fc911fa2d0ae60e230aac/pkg/upgrade/upgrades/upgrades.go#L197)):*
> ```Go
> 
> 	upgrade.NewTenantUpgrade(
> 		"only use session based leases",
> ```

nit: mention that it mimics `experimental_use_session_based_leasing=session`
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYjziV1wU0CbL2r3je:-NqYjziV1wU0CbL2r3jf:b2v7bsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L32)):*
> ```Go
> 	ctx context.Context, version clusterversion.ClusterVersion, deps upgrade.TenantDeps,
> ) error {
> 	return nil
> ```

should we fill this out in this commit too? i'm not actually sure what's needed.
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYhoEN3zfcVPPF_wEN:-NqYhoEN3zfcVPPF_wEO:b-y15q71) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L90)):*
> ```Go
> 	expected AS e, current AS c;
> `
> 	r := retry.StartWithCtx(ctx, retry.Options{})
> ```

nit: what are the default retry options? it might be better to make this explicit
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYi8Gg0GEYzMWaC6UB:-NqYi8Gg0GEYzMWaC6UC:b59824p) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L92)):*
> ```Go
> 	r := retry.StartWithCtx(ctx, retry.Options{})
> 	for r.Next() {
> 		row, err := deps.InternalExecutor.QueryRow(ctx,
> ```

nit: should it use QueryRowEx and specify the `node` user?
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYiWLRBAvWoMJSk34h:-NqYiWLRBAvWoMJSk34i:b-bkwnqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L103)):*
> ```Go
> 			return errors.AssertionFailedf("unexpected row from session based leasing drain query %v", row)
> 		}
> 		sessionBasedOnly := row[0].(*tree.DBool)
> ```

nit: use tree.MustBeDBool
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYipaQ-dqFxmPwl8ll:-NqYipaQ-dqFxmPwl8lm:b-9bp36) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L131)):*
> ```Go
> 		leaseTable.Families = newLeaseTableFormat.TableDesc().Families
> 		leaseTable.NextColumnID = newLeaseTableFormat.TableDescriptor.GetNextColumnID()
> 		leaseTable.NextIndexID = newLeaseTableFormat.TableDescriptor.GetNextIndexID()
> ```

looks fine, but is there anything we can do to make sure that these are _all_ of the descriptor fields that need updating?
___
*[`pkg/sql/catalog/lease/lease_internal_test.go` line 1585 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYUOcg6L1fLhJrDHRH:-NqYUOcg6L1fLhJrDHRI:brs8bhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e608c0a59d612a122f2fc911fa2d0ae60e230aac/pkg/sql/catalog/lease/lease_internal_test.go#L1585)):*
> ```Go
> 	// Validate all settings can be set and the provider works correctly.
> 	for idx, setting := range []string{"off", "dual_write", "drain", "session"} {
> 		sessionMode := SessionBasedLeasingMode(idx + 1)
> ```

nit: add a comment saying that we offset by 1 since "auto" is the first one. even better would be to add a function or a ByName map that converts string->`mode`
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease_test.go` line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYjX9CEGwyyk5Vqhhq:-NqYjX9CEGwyyk5Vqhhr:b-tbfb5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease_test.go#L102)):*
> ```Go
> 
> 	// Validate the lease table was converted properly.
> 	runner.Exec(t, "SELECT session_id from system.lease")
> ```

nit: should this check that the session ID is not null? we also might be able to compare it to the actual session ID

File: pkg/cli/zip_table_registry.go [hunk 1]
```diff
@@ -1100,12 +1100,9 @@ var zipSystemTables = DebugZipTableRegistry{
...
 	"system.lease": {
-		nonSensitiveCols: NonSensitiveColumns{
-			`"descID"`,
-			"version",
-			`"nodeID"`,
-			"expiration",
-		},
+		customQueryRedacted: `
+			SELECT * FROM system.lease;
+		`,
```
Comment (abarganier) [inline pkg/cli/zip_table_registry.go:1105]: > 		SELECT * FROM system.lease;

nit: I'm a bit concerned that using `SELECT *` here could cause us to leak newly added sensitive columns into redacted debug zip bundles. Any reason why we can't explicitly define the non-sensitive columns?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119041)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @rafiss)

___
*[`pkg/cli/zip_table_registry.go` line 1104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYe6II7M717sPGlDGp:-NqhkPH305kzTc_G0V6k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/cli/zip_table_registry.go#L1104)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

don't we want `customQueryUnredacted` here? nothing in the table should be sensitive data
</blockquote></details>

Done.
___
*[`pkg/cli/zip_table_registry.go` line 1105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119041#gh-1491223758:-NqhkQ-s3qn0kzZ1zgml:b-flmc3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a80f077f42e57cb9a0a5afd31a5b707ba1d7b/pkg/cli/zip_table_registry.go#L1105)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

> 		SELECT * FROM system.lease;

nit: I'm a bit concerned that using `SELECT *` here could cause us to leak newly added sensitive columns into redacted debug zip bundles. Any reason why we can't explicitly define the non-sensitive columns? 
</blockquote></details>

The problem is that we renamed columns and we need to support the mixed version state. The system.leases table will never sensitive / PII data because its focused on just tracking which versions of descriptors are in use.
___
*[`pkg/sql/catalog/lease/count.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYez2U6ISLrdg4T2dw:-Nqhn3rp0qcsjxDJ7f5K:b594kge) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/sql/catalog/lease/count.go#L44)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

`useSyntheticLeaseDescriptor` is a bit confusing, since below we are adding to the syntheticDescriptors slice even when `useSyntheticLeaseDescriptor` is false.

maybe we should name this bool and the function something like `useSessionBasedLeaseDescriptor`
</blockquote></details>

I renamed it to:
// Indicates if the leasing descriptor has been upgraded for session based  
// leasing.  
leasingDescIsSessionBased := hasSessionBasedLeasingDesc(ctx, settings)


It makes the usage hopefully more clear and added some comments
___
*[`pkg/sql/catalog/lease/lease.go` line 1638 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYguVp1Nr4oXdZptDK:-Nqi-23Q4158kPTog_Ri:b-vgfhjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/sql/catalog/lease/lease.go#L1638)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: the comment seems old now. should we switch to using DELETE WHERE, since we aren't using AOST now?
</blockquote></details>

We are doing AOST with a fixed timestamp below
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYjziV1wU0CbL2r3je:-NqjNewKAEPOOSQLFaE4:bm6ox4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L32)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we fill this out in this commit too? i'm not actually sure what's needed.
</blockquote></details>

Ugh, I should have just dropped this one.
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYhoEN3zfcVPPF_wEN:-NqiF2SZ-EFQVTXs2OMB:bv3zynt) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L90)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: what are the default retry options? it might be better to make this explicit
</blockquote></details>

Made this explicit by starting the retry half a second later and backing off till we are checking every 10 seconds.
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYipaQ-dqFxmPwl8ll:-NqiP2sn4V8PLvsoi_qV:bv091x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease.go#L131)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

looks fine, but is there anything we can do to make sure that these are _all_ of the descriptor fields that need updating?
</blockquote></details>

The upgrade test is modified to compare descriptors to make sure they match after minus version and modification time.
___
*[`pkg/upgrade/upgrades/v24_1_session_based_lease_test.go` line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-NqYjX9CEGwyyk5Vqhhq:-NqiV3NG-czkIcKsXkcf:b-gmggit) ([raw file](https://github.com/cockroachdb/cockroach/blob/98acdafcd52c26125d2d7a88bef622654166c499/pkg/upgrade/upgrades/v24_1_session_based_lease_test.go#L102)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should this check that the session ID is not null? we also might be able to compare it to the actual session ID
</blockquote></details>

Good point, and I added extra queries to confirm this.

**Review by rafiss — APPROVED** [review]
lgtm! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119041#-:-NqjcYbe0V3YbLVDJoD-:b-nzoors)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @fqazi)

**Comment by fqazi** [comment]
@rafiss TFTR!

bors r+

**Pull request merged**